File: week1/alfazya021/238-ProductofArrayExceptSelf.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] dp = new int[nums.length];
        dp[0] = 1;
        
        for(int i = 1; i < nums.length; i++) {
            dp[i] = dp[i-1] * nums[i - 1]; 
        }
        
        int right = 1;
        
        for(int i = nums.length - 1; i >= 0; i--) {
            dp[i] *= right;
            right *= nums[i];
        }
        
        return dp;
    }
}

File: week1/atakantekatan/238. Product of Array Except Self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prod = 1
        cnt = {}
        ans = []
        for n in nums:
            if n not in cnt:
                cnt.update({n:1})
            else:
                cnt[n] += 1
            if n!=0:
                prod *= n
        if 0 not in cnt:
            for num in nums:
                ans.append(prod//num)
        elif cnt[0]==1:
            for num in nums:
                if num == 0:
                    ans.append(prod)
                else:
                    ans.append(0)
        elif cnt[0]>1:
            ans = [0]*len(nums)
        return ans